File: Vandame/matlab_code/Btime_frequency_extraction.m
clear Data.spectroinfo
global Data
disp('spectrogram extraction');
Data.spectroinfo.overlap=0.5;

Data.spectroinfo.NBpoint_fft = 1024;

[~,Data.spectroinfo.F,Data.spectroinfo.T,~]=...
    spectrogram(squeeze(Data.directory(1).rawdata(1,:).'),hamming(Data.spectroinfo.NBpoint_fft),...
    floor(Data.spectroinfo.NBpoint_fft*Data.spectroinfo.overlap)+1,...
    Data.spectroinfo.NBpoint_fft,Data.wavinfo.SampleRate,'yaxis');

[Data.spectroinfo.gridT,Data.spectroinfo.gridF]=meshgrid(Data.spectroinfo.T,Data.spectroinfo.F);

[dim1,dim2] = size(Data.spectroinfo.gridT);

for i=1:NB_folder
    Data.directory(i).spectro=zeros(NB_file,dim1,dim2);
    for j=1:NB_file
        [~,~,~,Data.directory(i).spectro(j,:,:)]=(...
            spectrogram(squeeze(Data.directory(i).rawdata(j,:).'),hamming(Data.spectroinfo.NBpoint_fft),...
            floor(Data.spectroinfo.NBpoint_fft*Data.spectroinfo.overlap)+1,...
            Data.spectroinfo.NBpoint_fft,Data.wavinfo.SampleRate,'yaxis'));
            Data.directory(i).spectro(find(Data.directory(i).spectro(j,:,:)<10^-20))=10^-20;
         Data.directory(i).spectro(j,:,:)=(abs(Data.directory(i).spectro(j,:,:)));
    end
end
    Data.spectroinfo.s = squeeze(fliplr(size(Data.directory(i).spectro(j,:,:))));
    Data.spectroinfo.s = fliplr(Data.spectroinfo.s(1:2));
clear dim1 dim2 i j
% %% display
% figure(1)
% subplot(2,2,1);
% hold on;
% mesh(Data.spectroinfo.gridT/60,Data.spectroinfo.gridF*10^-3,20*log10(squeeze(Data.directory(end).spectro(end,:,:))));
% xlabel('time (mins)');
% ylabel('frequency (kHz)');
% view(0,90)
% colorbar

File: Vandame/ChatGPT_MATtoPY/Btime_frequency_extraction.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import spectrogram

print('spectrogram extraction')

class SpectroInfo:
    def __init__(self):
        self.overlap = 0.5
        self.NBpoint_fft = 1024
        self.F = None
        self.T = None
        self.gridT = None
        self.gridF = None

class Data:
    spectroinfo = SpectroInfo()
    spectroinfo.overlap = 0.5
    spectroinfo.NBpoint_fft = 1024
    spectroinfo.F, spectroinfo.T, _, _ = spectrogram(
        np.squeeze(Data.directory[0].rawdata[0].T),
        window='hamming',
        nperseg=spectroinfo.NBpoint_fft,
        noverlap=int(np.floor(spectroinfo.NBpoint_fft * spectroinfo.overlap)) + 1,
        fs=Data.wavinfo.sample_rate,
        axis='y'
    )
    spectroinfo.gridT, spectroinfo.gridF = np.meshgrid(spectroinfo.T, spectroinfo.F)

    dim1, dim2 = Data.spectroinfo.gridT.shape

    for i in range(NB_folder):
        Data.directory[i].spectro = np.zeros((NB_file, dim1, dim2))
        for j in range(NB_file):
            _, _, _, Data.directory[i].spectro[j] = spectrogram(
                np.squeeze(Data.directory[i].rawdata[j].T),
                window='hamming',
                nperseg=spectroinfo.NBpoint_fft,
                noverlap=int(np.floor(spectroinfo.NBpoint_fft * spectroinfo.overlap)) + 1,
                fs=Data.wavinfo.sample_rate,
                axis='y'
            )
            Data.directory[i].spectro[j][Data.directory[i].spectro[j] < 10e-20] = 10e-20
            Data.directory[i].spectro[j] = np.abs(Data.directory[i].spectro[j])

    Data.spectroinfo.s = np.flipud(Data.directory[i].spectro[j].shape)
    Data.spectroinfo.s = np.flipud(Data.spectroinfo.s[:2])

dim1 = None
dim2 = None
i = None
j = None

# Display
# fig = plt.figure(1)
# ax = fig.add_subplot(2, 2, 1, projection='3d')
# ax.plot_surface(
#     Data.spectroinfo.gridT / 60,
#     Data.spectroinfo.gridF * 10 ** -3,
#     20 * np.log10(np.squeeze(Data.directory[-1].spectro[-1])),
#     cmap='viridis'
# )
# ax.set_xlabel('Time (mins)')
# ax.set_ylabel('Frequency (kHz)')
# ax.view_init(elev=90, azim=0)
# fig.colorbar(ax)

